src/main/java/com/conveyal/r5/analyst/cluster/TimeGrid.java

```diff
@@ -105,40 +99,33 @@ public void setTarget(int targetIndex, int[] pixelValues) {
         }
     }
 
-    // FIXME why don't we write this straight to a stream instead of buffering and doing address math?
-    // Do we really want to buffer this in memory?
-    // It seems like the main reason to use a byte array is to get little-endian byte order for JavaScript typed arrays.
-    // Could we use a Guava little-endian output stream?
-    public byte[] writeGrid() {
+    /**
+     * Write the grid to a LittleEndianDataOutputStream
+     */
+    public void writeGridToStream(LittleEndianDataOutputStream outputStream) throws IOException {
         int sizeInBytes = nValues * Integer.BYTES + HEADER_SIZE;
         LOG.info("Writing travel time surface of size {} kB", sizeInBytes / 1000);
-        ByteBuffer buffer = ByteBuffer.allocate(sizeInBytes);
-        buffer.order(ByteOrder.LITTLE_ENDIAN);
+
         // Write header
-        buffer.put(gridType.getBytes());
-        buffer.putInt(version);
-        buffer.putInt(zoom);
-        buffer.putInt(west);
-        buffer.putInt(north);
-        buffer.putInt(width);
-        buffer.putInt(height);
-        buffer.putInt(nValuesPerPixel);
-
-        //TODO wrap below in gzip output stream, once other changes are made
-
-        // Write values, delta coded TODO why are we delta-coding within pixels?
-        int i = 0;
-        while (i < nValues) {
-            int prev = 0;
-            for (int v = 0; v < nValuesPerPixel; v++, i++) {
-                int curr = values[i];
+        outputStream.write(gridType.getBytes());
+        outputStream.writeInt(version);
+        outputStream.writeInt(zoom);
+        outputStream.writeInt(west);
+        outputStream.writeInt(north);
+        outputStream.writeInt(width);
+        outputStream.writeInt(height);
+        outputStream.writeInt(nValuesPerPixel);
+
+        // Write values, delta coded
```

Comment:
This nested loop could use some clearer comments, to the effect of: If we have several different values per pixel (unually N different percentiles), we write out one flattened 2D array for the first percentile, then one flattened 2D array for the second plane and so on, delta-coding within each of these flattened 2D arrays, on the principle that neighboring pixels have similar values within a given percentile.